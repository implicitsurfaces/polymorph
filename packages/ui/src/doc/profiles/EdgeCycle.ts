import { Halfedge, HalfedgeLike } from "./Halfedge";

export type EdgeCycleLike = HalfedgeLike[] | EdgeCycle;

/**
 * An edge cycle stores a sequence of consecutive halfedges that
 * starts and ends at the same point.
 *
 * The following invariants are enforced, where `n = cycle.halfedges.length`:
 * - `n > 0`
 * - For all i in [0..n-1]:
 *   `cycle.halfedges[i % n].endPoint() == cycle.halfedges[(i+1) % n].startPoint()`
 */
export class EdgeCycle {
  readonly halfedges: readonly Halfedge[];

  constructor(other: EdgeCycleLike) {
    if (other instanceof EdgeCycle) {
      this.halfedges = other.halfedges;
    } else {
      // Convert from from HalfedgeLike to Halfedge
      this.halfedges = other.map((h) => new Halfedge(h));

      // Check preconditions
      if (this.halfedges.length === 0) {
        throw Error(`An edge cycle must contain at least one edge.`);
      }
      let prevEndPoint = this.halfedges[this.halfedges.length - 1].endPoint();
      for (const halfedge of this.halfedges) {
        if (halfedge.startPoint() !== prevEndPoint) {
          throw Error(
            `Consecutive halfedges in an edge cycle must share their endpoint.`,
          );
        }
        prevEndPoint = halfedge.endPoint();
      }
    }
  }
}
