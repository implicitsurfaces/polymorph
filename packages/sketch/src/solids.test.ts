import { test } from "vitest";
import { ellipseConic } from "./conic";
import { asVec3, XY_PLANE, XZ_PLANE } from "./geom-3d";
import { SolidSlice } from "./profiles";
import { ConicFrustum } from "./solids";
import { expectASCIIDistance } from "./test-utils";
import { asNum } from "./num";

test("frustum from an ellipse conic", async () => {
  const ellipse = ellipseConic(asNum(0.8), asNum(0.2));
  const frustum = new ConicFrustum(ellipse, asNum(1));

  const p = XZ_PLANE.translate(asVec3(0, 0, 0.1));
  (await expectASCIIDistance(new SolidSlice(frustum, p))).toMatchSnapshot();

  const p2 = XY_PLANE.translate(asVec3(0, 0, 1));
  (await expectASCIIDistance(new SolidSlice(frustum, p2))).toMatchSnapshot();
});
