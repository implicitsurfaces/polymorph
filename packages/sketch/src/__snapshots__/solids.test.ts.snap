// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`frustum from an ellipse conic 1`] = `
"░░░░░░˖˖˖˖˖˖˖˖--------○○○-○◎◎◎◎◎◎◉◉◉◉-◉●●●●●-●●●--●--●●●-●●●●●◉-◉◉◉◉◎◎◎◎◎◎○-○○○--------˖˖˖˖˖˖˖˖░░░░░
░░░░░░░░˖˖˖˖˖˖˖˖-------○○○○○◎◎◎◎◎◎◉-◉◉◉●●-●●●●●-●●●●●-●●●●●-●●◉◉◉-◉◎◎◎◎◎◎○○○○○-------˖˖˖˖˖˖˖˖░░░░░░░
▒˖░░░˖░░░░˖˖˖˖˖˖˖˖-------○○○○○◎-◎--◉-◉◉◉●-●-●---●●●●●---●-●-●◉◉◉-◉--◎-◎○○○○○-------˖˖˖˖˖˖˖˖░░░░˖░░░˖
˖˖˖˖˖░░░░░░░˖˖˖˖˖˖˖------○○○○○-◎◎◎◎◎◉◉◉-◉●-●●●-●-●-●-●-●●●-●◉-◉◉◉◎◎◎◎◎-○○○○○------˖˖˖˖˖˖˖░░░░░░░˖˖˖˖
˖▒▒▒˖˖˖░░░░░░░˖˖˖˖˖˖˖------○-○-○◎◎--◎◉--◉-●-●-●●●●-●●●●-●-●-◉--◉◎--◎◎○-○-○------˖˖˖˖˖˖˖░░░░░░░˖˖˖▒▒▒
▒▒˖˖▒▒˖-▒░░˖˖░░░˖˖˖˖˖˖------○----◎◎-◎◉-◉◉◉●●●●-●●●●●●●-●●●●◉◉◉-◉◎-◎◎----○------˖˖˖˖˖˖░░░˖˖░░▒-˖▒▒˖˖▒
▓˖▒▒˖˖▒▒▒˖▒░░░░˖░˖˖˖˖˖˖˖-----○○○--◎◎--◉---◉●●●-●●●●●●●-●●●◉---◉--◎◎--○○○-----˖˖˖˖˖˖˖░˖░░░░▒˖▒▒▒˖˖▒▒˖
▓▓˖▓▓▒▒▒▒-▒˖˖░░˖░˖˖˖˖˖˖˖˖˖-----○-○○◎-◎--◉---●●●●●●-●●●●●●---◉--◎-◎○○-○-----˖˖˖˖˖˖˖˖˖░˖░░˖˖▒-▒▒▒▒▓▓˖▓
▓˖▓▓▓▓▓▒˖˖▒▒▒▒˖▒░░░░░░˖˖˖˖˖-----○-○○◎◎◎-◉◉◉◉--●●●-●-●●●--◉◉◉◉-◎◎◎○○-○-----˖˖˖˖˖░░░░░░▒˖▒▒▒▒˖˖▒▓▓▓▓▓˖
▓▓▓▓▓▓▓▓▓˖▒˖▒▒▒▒˖˖░░░░░░˖˖˖˖˖----○○○○◎◎◎◎◉◉◉◉-◉●-●-●-●◉-◉◉◉◉◎◎◎◎○○○○----˖˖˖˖˖░░░░░░˖˖▒▒▒▒˖▒˖▓▓▓▓▓▓▓▓
███▓▓▓▓▓▓▓˖▓▓▒▒▒▒▒-˖░░░˖░░˖˖˖˖-----○○○◎◎◎◎◉◉◉◉◉◉●---●◉◉◉◉◉◉◎◎◎◎○○○-----˖˖˖˖░░˖░░░˖-▒▒▒▒▒▓▓˖▓▓▓▓▓▓▓██
██˖██˖█˖˖▓▓˖▓▓▓▓▒▒▒▒▒˖˖˖˖˖░░˖˖˖˖----○○○-◎◎-◉◉◉◉◉-◉-◉-◉◉◉◉◉-◎◎-○○○----˖˖˖˖░░˖˖˖˖˖▒▒▒▒▒▓▓▓▓˖▓▓˖˖█˖██˖█
███████˖██▓▓▓▓▓˖▓▓▓▒▒▒▒˖˖˖˖˖˖˖˖˖˖˖----○○◎-◎◎--◉◉-◉●◉-◉◉--◎◎-◎○○----˖˖˖˖˖˖˖˖˖˖˖▒▒▒▒▓▓▓˖▓▓▓▓▓██˖██████
█˖███████˖██˖▓▓▓▓▓▓▓▓▒▒▒▒▒˖░░░░░˖˖˖---○○-◎◎◎◎◎◉--◉◉◉--◉◎◎◎◎◎-○○---˖˖˖░░░░░˖▒▒▒▒▒▓▓▓▓▓▓▓▓˖██˖███████˖
˖█████˖██˖˖█████˖▓▓▓▓▓▓▓▒▒▒▒▒▒░░░░˖˖˖---○○◎◎--◎◎◉◉-◉◉◎◎--◎◎○○---˖˖˖░░░░▒▒▒▒▒▒▓▓▓▓▓▓▓˖█████˖˖██˖█████
˖██████████████████▓▓▓˖▓▓▓▒▒▒▒▒▒░░░░˖˖-----○◎◎-◎◎-◉-◎◎-◎◎○-----˖˖░░░░▒▒▒▒▒▒▓▓▓˖▓▓▓██████████████████
˖█████˖███████████˖˖█▓▓˖▓˖▓▓˖˖▒▒▒˖░░░░˖˖---○○◎◎◎◎◎-◎◎◎◎◎○○---˖˖░░░░˖▒▒▒˖˖▓▓˖▓˖▓▓█˖˖███████████˖█████
██˖██˖██˖˖█████████˖████▓▓˖˖▓˖˖˖˖▒▒▒˖░░░˖˖-○-○○-◎---◎-○○-○-˖˖░░░˖▒▒▒˖˖˖˖▓˖˖▓▓████˖█████████˖˖██˖██˖█
█████████████˖█˖██████████˖▓▓▓▓▓▓˖▒˖▒▒░░░░˖--○○○-○-○-○○○--˖░░░░▒▒˖▒˖▓▓▓▓▓▓˖██████████˖█˖████████████
███˖███˖████˖████████˖█████▓▓▓▓▓▓▓▒▒▒▒▒˖░░░˖˖-○○○○○○○○○-˖˖░░░˖▒▒▒▒▒▓▓▓▓▓▓▓█████˖████████˖████˖███˖██
███˖████████████████████████˖˖▓▓▓▓˖▒▒▒▒▒▒░˖░░˖--○-○-○--˖░░˖░▒▒▒▒▒▒˖▓▓▓▓˖˖████████████████████████˖██
██˖█████████████████████˖˖██˖˖▓▓˖▓▓▒▒˖˖▒░░░░░░˖˖-----˖˖░░░░░░▒˖˖▒▒▓▓˖▓▓˖˖██˖˖█████████████████████˖█
˖█˖˖█████████˖██˖█████˖██████▓▓▓▓˖▓▒▒▒˖▒░░░░˖░˖˖˖˖-˖˖˖˖░˖░░░░▒˖▒▒▒▓˖▓▓▓▓██████˖█████˖██˖█████████˖˖█
˖█˖˖˖████████˖██˖████˖███████▓▓▓▓˖˖▒▒˖▒▒░░░░░░˖˖˖˖-˖˖˖˖░░░░░░▒▒˖▒▒˖˖▓▓▓▓███████˖████˖██˖████████˖˖˖█
██˖███████████████████˖█˖˖███˖▓▓˖▓▓▒▒▒˖▒░░░░░░˖˖-----˖˖░░░░░░▒˖▒▒▒▓▓˖▓▓˖███˖˖█˖███████████████████˖█
██████˖█████████████████████˖˖˖▓▓▓˖▒▒▒▒▒▒░░░░˖---○○○---˖░░░░▒▒▒▒▒▒˖▓▓▓˖˖˖█████████████████████˖█████
██˖˖███˖███˖˖████████˖█˖██˖▓▓▓▓▓▓▓˖▒▒▒▒▒░░░˖˖-○○○○○○○○○-˖˖░░░▒▒▒▒▒˖▓▓▓▓▓▓▓˖██˖█˖████████˖˖███˖███˖˖█
˖██████████████˖██████˖███▓˖▓▓▓▓▓▒▒▒▒▒˖░░░˖--○○○○---○○○○--˖░░░˖▒▒▒▒▒▓▓▓▓▓˖▓███˖██████˖██████████████
████████˖˖███████████˖██▓▓˖˖▓˖˖˖▒▒▒▒░░░░˖˖-○○○○◎◎---◎◎○○○○-˖˖░░░░▒▒▒▒˖˖˖▓˖˖▓▓██˖███████████˖˖███████
█˖█˖██˖█˖█████████˖˖█▓▓˖▓˖▓▓˖▒▒▒▒▒░░░░˖˖--○○○◎◎◎◎◎-◎◎◎◎◎○○○--˖˖░░░░▒▒▒▒▒˖▓▓˖▓˖▓▓█˖˖█████████˖█˖██˖█˖
███████████████████▓▓▓˖▓▓▓▒▒▒▒▒▒˖░░░˖˖----○-----◎◎◉◎◎-----○----˖˖░░░˖▒▒▒▒▒▒▓▓▓˖▓▓▓██████████████████
˖█████˖██˖˖█████˖▓▓▓▓▓˖▓▒▒˖▒˖▒░░░░˖˖˖---○-◎◎◎-◎◎◉◉-◉◉◎◎-◎◎◎-○---˖˖˖░░░░▒˖▒˖▒▒▓˖▓▓▓▓▓˖█████˖˖██˖█████
█˖███████˖███▓▓▓▓▓˖▓˖▒▒▒▒▒˖░░░░░˖˖˖---○○-◎◎◎◎◎◉-◉◉◉◉◉-◉◎◎◎◎◎-○○---˖˖˖░░░░░˖▒▒▒▒▒˖▓˖▓▓▓▓▓███˖███████˖
███˖˖█████▓▓▓▓▓▓˖▓˖▒˖˖▒˖˖˖░░░░˖˖˖˖----○○◎-◎◎-◉--◉◉●◉◉--◉-◎◎-◎○○----˖˖˖˖░░░░˖˖˖▒˖˖▒˖▓˖▓▓▓▓▓▓█████˖˖██
███˖˖˖█▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒░░░░░˖˖˖˖----○○○◎◎◎◎-----◉---◉-----◎◎◎◎○○○----˖˖˖˖░░░░░▒▒▒▒▒▒▒▓▓▓▓▓▓▓▓▓█˖˖˖██
███˖▓▓▓▓▓▓▓▓▓˖▒▒▒▒▒˖░░░˖░░˖˖˖˖----○○-○---◎◉◉◉◉◉◉●●-●●◉◉◉◉◉◉◎---○-○○----˖˖˖˖░░˖░░░˖▒▒▒▒▒˖▓▓▓▓▓▓▓▓▓˖██
▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒˖˖░░░˖░░˖˖˖˖˖----○○-○◎◎◎◎◉-◉◉--●●●-●●●--◉◉-◉◎◎◎◎○-○○----˖˖˖˖˖░░˖░░░˖˖▒▒▒▒▒▒▓▓▓▓▓▓▓▓▓
▓˖˖▓▓▓▓▒˖▒▒▒▒˖▒˖░░░░░˖˖˖˖˖˖-----○-○○◎◎◎----◉--●●●●-●●●●--◉----◎◎◎○○-○-----˖˖˖˖˖˖░░░░░˖▒˖▒▒▒▒˖▒▓▓▓▓˖˖
▓▓▓▓▓▒▒▒▒-▒˖˖░░░˖░˖░˖˖˖˖˖˖-----○○○--◎◎-◉◉◉-◉●●-●●---●●-●●◉-◉◉◉-◎◎--○○○-----˖˖˖˖˖˖░˖░˖░░░˖˖▒-▒▒▒▒▓▓▓▓
▓▓▒▒˖▒▒▒▒˖▒░˖░░░░˖˖˖˖˖˖˖-----○-○-○◎--◎◉◉-◉-●●●●●●●●●●●●●●●-◉-◉◉◎--◎○-○-○-----˖˖˖˖˖˖˖░░░░˖░▒˖▒▒▒▒˖▒▒▓
▒˖▒˖▒▒˖˖▒░░░░░░░˖˖˖˖˖˖˖-----○○-○○◎-◎◎◉◉◉◉◉●●-●●●●●-●●●●●-●●◉◉◉◉◉◎◎-◎○○-○○-----˖˖˖˖˖˖˖░░░░░░░▒˖˖▒▒˖▒˖
˖▒▒▒˖˖˖░░░░░░░˖˖˖˖˖˖˖------○-○-○-◎◎--◉◉◉◉◉-●--●●-●-●-●●--●-◉◉◉◉◉--◎◎-○-○-○------˖˖˖˖˖˖˖░░░░░░░˖˖˖▒▒▒
˖˖˖˖˖░░░░░░░˖˖˖˖˖˖˖------○-○-○-◎◎◎-◎◉◉◉◉-●-●●●●●-●-●-●●●●●-●-◉◉◉◉◎-◎◎◎-○-○-○------˖˖˖˖˖˖˖░░░░░░░˖˖˖˖
▒˖░░˖░░░░░˖˖˖˖˖˖˖˖------○-○○-○-◎-◎--◉◉-◉●●-●●●●-●●●●●-●●●●-●●◉-◉◉--◎-◎-○-○○-○------˖˖˖˖˖˖˖˖░░░░░˖░░˖
░░░˖░░░░˖˖˖˖˖˖˖˖-------○○○-○◎-◎◎◎◎-◉◉◉◉●●●●●----●●●●●----●●●●●◉◉◉◉-◎◎◎◎-◎○-○○○-------˖˖˖˖˖˖˖˖░░░░˖░░
░░░░░˖˖˖˖˖˖˖˖˖-------○○-○○-◎---◎◎-◉◉◉◉◉●●-●●●●●-●●●●●-●●●●●-●●◉◉◉◉◉-◎◎---◎-○○-○○-------˖˖˖˖˖˖˖˖˖░░░░
░░˖░˖˖˖˖˖˖˖˖˖-------○-○○--◎◎◎◎---◉◉◉◉◉●●●●●●●-●---●---●-●●●●●●●◉◉◉◉◉---◎◎◎◎--○○-○-------˖˖˖˖˖˖˖˖˖░˖░
░░░˖˖˖˖˖˖˖˖----------○○-○◎◎-◎---◉◉-◉-●●--●-●-●--●●-●●--●-●-●--●●-◉-◉◉---◎-◎◎○-○○----------˖˖˖˖˖˖˖˖░░
░˖˖˖˖˖˖˖˖˖------------○○◎◎◎◎-◎◎◉◉◉◉◉◉-●●-●-●-●●●-●●●-●●●-●-●-●●-◉◉◉◉◉◉◎◎-◎◎◎◎○○------------˖˖˖˖˖˖˖˖˖
˖˖˖˖˖˖˖˖--------○○○-○○-◎◎-◎◎◎◎◉◉-◉◉--●●●●●●●●●●●●●-●●●●●●●●●●●●●--◉◉-◉◉◎◎◎◎-◎◎-○○-○○○--------˖˖˖˖˖˖˖"
`;

exports[`frustum from an ellipse conic 2`] = `
"████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓████████████████████████████████████████████████
███▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓████████████████████████
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓███████████
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓██
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▓▓▓▓▓
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░░░░░░░░░░░░░░░▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▒▒▒▒▒▒▒▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▒▒▒▒▒▒▒▒▒
▒▒▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▒▒▒
░░░░░░░░░░░░░░░░░░░░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░░░░░░░░░░░░░░░░░░░░░
░░░░░░░░░░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░░░░░░░░░░░
░░░░░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖-˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░░░░░
░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖-------------------------------------------------˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░
░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖-----------------------------------------------------------------˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░
░░░˖˖˖˖˖˖˖˖˖˖˖----------------------○○○○○○○○○○○○○○-○○○○○○○○○○○○○-----------------------˖˖˖˖˖˖˖˖˖˖˖░░
░░˖˖˖˖˖˖˖˖˖---------------○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○---------------˖˖˖˖˖˖˖˖˖░
░░˖˖˖˖˖˖˖˖--------○-○---○◎-◎-◎◎◉◉◉◉◉◉●●-●●●●●●●●●●-●●●●●●●●●●-●●◉◉◉◉◉◉◎◎-◎-◎○---○-○--------˖˖˖˖˖˖˖˖░
░░˖˖˖˖˖˖˖˖˖---------------○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○---------------˖˖˖˖˖˖˖˖˖░
░░░˖˖˖˖˖˖˖˖˖˖˖-----------------------○○○○○○○○○○○○○-○○○○○○○○○○○○○○----------------------˖˖˖˖˖˖˖˖˖˖˖░░
░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖-----------------------------------------------------------------˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░
░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖-------------------------------------------------˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░
░░░░░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖-˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░░░░░
░░░░░░░░░░░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░░░░░░░░░░
░░░░░░░░░░░░░░░░░░░░░░░░░░˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖˖░░░░░░░░░░░░░░░░░░░░░░░░
▒▒▒▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▒▒
▒▒▒▒▒▒▒▒▒▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▒▒▒▒▒▒▒
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░░░░░░░░░░░░░░░▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▓▓▓▓▓▓
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
███▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
█████████████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓██
█████████████████████████████████████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓███████████████████████████
████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████████████████████████████████████████████████████████████████████████████"
`;
